### PR TITLE
Enable computing fast metrics as a standalone operation

### PR DESCRIPTION
**Pull request by iainrb**
- Runs metrics in the `fast_metric_finder` class for RNASeqQC
- New command-line script `write_fast_metrics.py`
- Shared functionality between "write fast metrics" and "write all metrics" refactored into new classes 
- Updated tests and documentation

### CODE REVIEW

File: bam_qc_metrics/bam_qc.py [hunk 1]
```diff
@@ -727,8 +812,69 @@ def run_samtools_stats(self, args, strict=False):
...
+    """
+    Wrapper for the fast_metric_finder class, to read parameters, set up logging, and write JSON.
+    Enables fast metrics to be computed as a standalone action (eg. for RNASeqQC).
+    """
+
+    def __init__(self, config):
+        self.validate_config_fields(config)
+        # set up logging
+        self.debug = config[self.CONFIG_KEY_DEBUG]
+        self.verbose = config[self.CONFIG_KEY_VERBOSE]
```
Comment (slazicoicr) [inline bam_qc_metrics/bam_qc.py]: `self.debug` and `self.verbose` are not used.

File: bin/run_bam_qc.py [hunk 2]
```diff
@@ -3,89 +3,37 @@
...
-        valid = validate_positive_integer(args.skip_below_mapq, 'Quality score')
+        valid = validator.validate_positive_integer(args.skip_below_mapq, 'Quality score')
     if args.insert_max != None:
-        valid = valid and validate_positive_integer(args.insert_max, 'Max insert size')
+        valid = valid and validator.validate_positive_integer(args.insert_max, 'Max insert size')
     if args.sample != None:
-        valid = valid and validate_positive_integer(args.sample, 'Downsampling level')
-        valid = validate_sample_level(args.all_reads, args.sample)
+        valid = valid and validator.validate_positive_integer(args.sample, 'Downsampling level')
+        valid = validator.validate_sample_level(args.all_reads, args.sample)
```
Comment (slazicoicr) [inline bin/run_bam_qc.py]: This needs to be `valid and validator...`

File: bin/write_fast_metrics.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python3
+
+"""Script to compute faster BAM QC metrics derived from samtools stats"""
+
+import argparse, cProfile, os, re, sys, tempfile
```
Comment (slazicoicr) [inline bin/write_fast_metrics.py]: Unused imports

**Comment by iainrb** [comment]
> 9 out of the 11 test cases fail on my machine.

That's unfortunate. I wonder why they passed for me. I'll get back to you on this.

File: bam_qc_metrics/bam_qc.py [hunk 4]
```diff
@@ -1,27 +1,146 @@
...
+            valid = False
+        elif int(sample_level) < MINIMUM_SAMPLE_LEVEL:
+            msg = "ERROR: Minimum sample level is %i reads." % MINIMUM_SAMPLE_LEVEL
+            msg = msg+" Increase the --sample argument, or use --all to omit downsampling.\n"
+            sys.stderr.write(msg)
+            valid = False
+        return valid
+
+
+class version_updater(base_constants):
```
[the remark this answers is not in the document]
Comment (iainrb) [reply bam_qc_metrics/bam_qc.py]: Yes it does. It uses PACKAGE_VERSION_KEY: https://github.com/oicr-gsi/bam-qc-metrics/blob/074869e3ea81f2d767c9cad4acfb486a098e2fdd/bam_qc_metrics/bam_qc.py#L156

Note as well that I have modified the class to fix a failing test (the only one which was failing for me): https://github.com/oicr-gsi/bam-qc-metrics/commit/4d5b1d30539f11125f41ecd49db050e1a9410203

Work on the other issues you identified is in progress. :-)

File: bam_qc_metrics/bam_qc.py (hunk 1 above)
Comment (iainrb) [reply bam_qc_metrics/bam_qc.py]: They are used by the following call to self.configure_logger. Although, as you rightly point out, that method needs to be modified so the base class doesn't impllicitly assume self.debug and self.verbose exist.

File: bin/run_bam_qc.py (hunk 2 above)
Comment (iainrb) [reply bin/run_bam_qc.py]: Yes, will fix

File: bin/write_fast_metrics.py (hunk 3 above)
Comment (iainrb) [reply bin/write_fast_metrics.py]: Yes, will fix

File: bam_qc_metrics/bam_qc.py [hunk 5]
```diff
@@ -1,27 +1,146 @@
...
+    """
+    Class for methods shared between bam_qc and fast_metric_writer
+    """
+
+    def configure_logger(self, log_path=None):
+        logger = logging.getLogger(__name__)
+        if self.debug:
+            log_level = logging.DEBUG
+        elif self.verbose:
+            log_level = logging.INFO
```
[the remark this answers is not in the document]
Comment (iainrb) [reply bam_qc_metrics/bam_qc.py]: I think that would be too complicated. Instead I am making `debug` and `verbose` optional arguments to `configure_logger`. Which anyway is neater than creating instance variables which are never used again after the logger is configured.

File: bam_qc_metrics/bam_qc.py [hunk 6]
```diff
@@ -50,20 +168,14 @@ def update_files(self):
 
 class bam_qc(base):
 
-    CONFIG_KEY_BAM = 'bam'
-    CONFIG_KEY_DEBUG = 'debug'
-    CONFIG_KEY_INSERT_MAX = 'insert max'
-    CONFIG_KEY_LOG = 'log path'
     CONFIG_KEY_MARK_DUPLICATES = 'mark duplicates'
     CONFIG_KEY_METADATA = 'metadata'
     CONFIG_KEY_N_AS_MISMATCH = 'n as mismatch'
```
[the remark this answers is not in the document]
Comment (iainrb) [reply bam_qc_metrics/bam_qc.py]: Will fix

File: bam_qc_metrics/bam_qc.py [hunk 7]
```diff
@@ -1,27 +1,146 @@
...
+        return valid
+
+    @staticmethod
+    def validate_sample_level(sample_all, sample_level):
+        # assumes sample_level is a positive integer (not None)
+        valid = True
+        if sample_all == True:
+            sys.stderr.write("ERROR: Cannot specify both --all-reads and --sample\n")
+            valid = False
+        elif int(sample_level) < MINIMUM_SAMPLE_LEVEL:
```
[the remark this answers is not in the document]
Comment (iainrb) [reply bam_qc_metrics/bam_qc.py]: Yes, will fix

File: bam_qc_metrics/bam_qc.py [hunk 8]
```diff
@@ -1,27 +1,146 @@
...
+    def configure_logger(self, log_path=None):
+        logger = logging.getLogger(__name__)
+        if self.debug:
+            log_level = logging.DEBUG
+        elif self.verbose:
+            log_level = logging.INFO
+        else:
+            log_level = logging.WARN
+        logger.setLevel(log_level)
+        handler = None
```
[the remark this answers is not in the document]
Comment (iainrb) [reply bam_qc_metrics/bam_qc.py]: I'm not entirely sure what you mean by this? I have removed the `else` clause and set `log_level = logging.WARN` as a default before the `if/elif` is evaluated.

File: bam_qc_metrics/bam_qc.py (hunk 1 above)
Comment (iainrb) [reply bam_qc_metrics/bam_qc.py]: `debug` and `verbose` are no longer instance variables, instead they are simply arguments to `configure_logger()`

**Pull request merged**